Rewrite (thereby fix) types for express-session (second attempt)

HoldYourWaffle opened the pull request:
# This is an updated version of #41809

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: The changes of the first attempt (#41809) have been discussed and reviewed by the package maintainer in https://github.com/expressjs/session/pull/656.
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.


### This PR is a complete rewrite, because the current definitions are simply wrong
The exact changelog (including some of the reasons why the current definitions are wrong) is listed in https://github.com/expressjs/session/pull/656.
The original package maintainer @dougwilson has already reviewed the first-attempt version of these definitions and confirmed that they were correct. @dougwilson, could you check/confirm if these are still accurate?

Although I haven't thoroughly tested it, I'd be very surprised if this rewrite would not be a breaking change (at least concerning import syntax). However, these fixes are desperately needed because the current definitions are pretty much unusable.
This issue has also been noticed in https://github.com/JLuboff/connect-mssql-v2/issues/10.

Unfortunately this fix will most likely break downstream definitions. I have already [comitted to updating the definitions for `express-mysql-session`](https://github.com/chill117/express-mysql-session/pull/93) because I'm using it myself, but the other session stores will (probably) need to be updated as well. I noticed @JLuboff and @bradtaniguchi were also having troubles with the current definitions, perhaps they can shine some light on how their session store would be affected by this?


### As for the definitions themselves
When I wrote these definitions (more than a year ago now), I had to declare the `SessionData` interface in the global scope to make declaration merging work. This feels wrong though, was this (and is this still) actually required or did I miss something?

In the long run I'd prefer integrating these definitions into `express-session` itself, but as discussed in https://github.com/expressjs/session/pull/656 this'll have to wait until the v2 release.

This PR (temporarily) closes #35194 and almost certainly fixes #18529, #18527 and #16480.

HoldYourWaffle commented:
Ah shit, here we go again...

HoldYourWaffle commented:
Shhh bot, I'm working on it...

HoldYourWaffle commented:
Hold on, how am I supposed to fix this error:
```
Error in passport.socketio
Error: Errors in typescript@4.0 for external dependencies:
##[error]node_modules/connect-mongo/src/types.d.ts(113,66): error TS2694: Namespace 'global.Express' has no exported member 'SessionData'.
##[error]node_modules/connect-mongo/src/types.d.ts(114,45): error TS2694: Namespace 'global.Express' has no exported member 'SessionData'.
##[error]node_modules/connect-mongo/src/types.d.ts(118,47): error TS2694: Namespace 'global.Express' has no exported member 'SessionData'.
```

This is caused by the `connect-mongo` typings being out-of-date due to this PR. So I guess I'm supposed to update the types of `connect-mongo`, *which kindof makes sense*, except for the fact that I can't update those typings because they are only compatible with these new definitions, which can't be merged, because they are not compatible with the old definitions.

So what am I supposed to do here?

HoldYourWaffle commented:
@peterblazejewicz Good question, I honestly have no idea. I initially wrote these definitions over a year ago, could just have been the force of habit.

Would something like this more accurately represent `expres-session`'s behavior?
```ts
export = session;

declare function session(options?: session.SessionOptions): express.RequestHandler;

declare namespace session {
    export interface ... { ... }
}
```

peterblazejewicz commented:
Yes, it would, thanks!

HoldYourWaffle commented:
Alright then, I'll change that.

How do I properly import with this syntax though? Without `esModuleInterop` I'm only allowed to do `import session = require('express-session')`, which is fine for the "default function reference", but it gets really ugly for 'type references':

```ts
import session = require('express-session');

app.use(session({ // this is fine
    store: new session.MemoryStore(), // this is ugly
    ...
}));
```

I've never fully understood these weird import/export/module/interoperability shenanigans, so I'm sorry if this is a dumb question 😅

peterblazejewicz commented:
> How do I properly import with this syntax though?

I think it could be:
```ts
import session = require('express-session');
import { MemoryStore } from 'express-session';
```

This is for DT only, consumers of the package can use 'esModuleInterop' as you noticed (though it will be further improved in newer version of TS for better code reuse AFAIK).

HoldYourWaffle commented:
Ah crap, this means I'll have to revert most of my hard manual updating...

HoldYourWaffle commented:
As far as I can see there are three errors left now:

1. The [circular error with external dependencies](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46576#issuecomment-669489207)
2. Declaration merging is broken for `easy-session`, I'm too tired to fix that right now so I'll take another shot at that tomorrow.
3. The declaration of `SessionData` in the global scope is not so much an *error*, but it's *undesired*. I'll see if I can fix that while preserving declaration merging tomorrow as well.

HoldYourWaffle commented:
Looks like NPM is having server issues, so that's nice.

HoldYourWaffle commented:
To increase my developer-citizen-karma and rerun the tests that failed due to NPM's issues, I have copy-pasted (and reformatted) the documentation from `express-session`'s README into the type definitions as tsdoc.

Unfortunately I couldn't find anything on `CookieOptions.signed`, `CookieOptions.encode`, `Store.regenerate`, `Store.load` and `Store.createSession`. I'll see if I can find anything on those later today.

HoldYourWaffle commented:
Alright, so the only issues remaining are the [external dependencies](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46576#issuecomment-669489207), but I have no idea what I'm supposed to do with that.

Am I supposed to convince `connect-mongo` to merge broken type definitions?

Should I open a PR for `connect-mongo`, temporarily change the dependency reference here to my fork, merge this PR, merge `connect-mongo`'s PR and then finally open another PR here to fix the dependency reference?

@RyanCavanaugh, can I use you to "request human assistance"? Your bot doesn't appear to have a way to do that unfortunately...

HoldYourWaffle commented:
This is a mess. I apologize to everyone that's (still) getting notified for each CI failure...
I hope you're only getting a notification for the bot's comment and not a separate one for the CI failure as well like me...

danger-public commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-Danger;
-->



Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## easy-session ([<kbd>unpkg</kbd>](https://unpkg.com/browse/easy-session@latest/))
was missing the following properties:
1. can
## express-session ([<kbd>unpkg</kbd>](https://unpkg.com/browse/express-session@latest/))
was missing the following properties:
1. Session
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against fb09491294e91f9046528f667bc97c0823d96ab5
</p>

HoldYourWaffle commented:
It's a miracle, CI tests pass! (Albeit with a hacky and temporary solution for connect-mongo...)

The danger bot is right though, turns out `Session` is a class rather than an interface.
However, if I change that, and have `Session` implement `SessionData`, TS complains about missing properties when you use declaration merging to add properties to `SessionData`.

Is there a way to tell TypeScript: "Hey, I know I'm missing properties here, but those will be magically added by JavaScript code"? Something like a [definite assignment assertion](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#strict-class-initialization) but for interface implementation.
I unfortunately can't use a `@ts-ignore` comment because they're banned and still break declaration merging.

kstiehl commented:
I agree Session should be a class. But I am not quite sure whether `Session` should implement `SessionData`. 
You should try to use `Session & SessionData` in some places. 

So for example 

     namespace Express {
        // Inject additional properties on express.Request
        interface Request {
            /**
             * This request's `Session` object.
             * Even though this property isn't marked as optional, it won't exist until you use the `express-session` middleware
             * [Declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) can be used to add your own properties.
             *
             * @see SessionData
             */
            session: session.Session & session.SessionData;

            /**
             * This request's session ID.
             * Even though this property isn't marked as optional, it won't exist until you use the `express-session` middleware
             */
            sessionID: string;

Using intersection types is quite common and I think its the perfect fit here.

https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html

HoldYourWaffle commented:
That's an interesting take I hadn't considered before. It makes sense, however,

> an object of class Session does not generally satisfy the SessionData interface

I'm not sure if this is true, given the following code in `Session`'s constructor ([source](https://github.com/expressjs/session/blob/4715f3b4ed75f7a3926137f6754cde42b9fb66b7/session/session.js#L29-L34)):
```ts
// merge data into this, ignoring prototype properties
for (var prop in data) {
    if (!(prop in this)) {
      this[prop] = data[prop]
    }
}
```

The way I interpret this, a `Session` object will always satisfy the `SessionData` interface, because all properties of a `SessionData` instance are merged into the `Session`.

Using an intersection type could definitely *work*, but I feel like it could lead to mistakes and confusion if people (mistakenly) use the types without intersection. My first intuition was to do something like this (untested 'pseudo' code):
```ts
class SessionHandler { ... }
interface SessionData { ... }

interface Session extends SessionHandler, SessionData { }

// Injection
interface Request {
    session: Session;
}
```
But this doesn't work, since we can't "just rename" the `Session` class.

Don't get me wrong, using intersection types would definitely be a good "workaround" if there's no better option. I'm just not sure if they're the most accurate and 'convenient' way to model this.

kstiehl commented:
> The way I interpret this, a Session object will always satisfy the SessionData interface, because all properties of a SessionData instance are merged into the Session.

That's true if the data parameter is set. Which is not the case for a newly created Session as you can see [here](https://github.com/expressjs/session/blob/4715f3b4ed75f7a3926137f6754cde42b9fb66b7/index.js#L160).

It doesn't feel like a workaround to me it might be unusual compared to other object-oriented programming languages but it's a widely used feature in typescript. 

However to propose another Idea which would also work is using Mapped types. To allow a partial implementation of `SessionData`. See [here](https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types) 

So you could do something like this 

```
type Partial<T> = {
  [P in keyof T]?: T[P];
};
```

Which would allow you to do something like that ...

```
interface Session extends SessionHandler, Partial<SessionData> { }
```

The `?` in the `Partial` type maps every property of `SessionData` to an optional one. So `Session` would always satisfy the interface. The autocompletion of your IDE would show you the properties of the `SessionData` interface but like I said they would always be optional on a Session instance. 
Which is technically correct because in a newly created session there will never be any property from `SessionData` just in the subsequent request


In the end, it's probably a matter of taste.  :smile:

HoldYourWaffle commented:
> That's true if the data parameter is set. Which is not the case for a newly created Session as you can see here.

Ahh, that makes sense. In that case my argument is invalid.

> It doesn't feel like a workaround to me it might be unusual compared to other object-oriented programming languages but it's a widely used feature in typescript.

Oh don't get me wrong, I'm definitely not *against intersection types*, they're a perfectly valid and as you say widely used feature in TypeScript. The reason I labeled the approach as a "workaround" was that I was still in the assumption that a `Session` object *always* satisfied the `SessionData` interface, in which case an intersection-"relationship" wouldn't make much sense.

<br>

> The autocompletion of your IDE would show you the properties of the SessionData interface but like I said they would always be optional on a Session instance.
Which is technically correct because in a newly created session there will never be any property from SessionData just in the subsequent request

This is something I'm not a fan of though.
It's a similar issue to `session` and `sessionID` on `Request`. I marked these as non-optional, even though they only exist after the `express-session` middleware has been used. I did this because those properties will always exist if the module is used properly, sparing the need for unnecessary (and possibly confusing) type assertions.

Similarly, if we declare `class Session implements Partial<SessionData>`, the user will *always* have to do "useless" type assertions on `SessionData` properties, even though they are not really optional, **at least as long as the session is properly initialized**. It's a trade-off between being "*technically* more accurate" and being "easier to use" for people that know what they're doing.
However, I recognize that this trade-off is much more 'fuzzy' in this situation than with the `Request` properties, and I'm happy to debate this stance.

By the way, `Partial` is a [built-in type](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html#partial-readonly-record-and-pick), so we wouldn't have to define it ourselves ;)

kstiehl commented:
>By the way, Partial is a built-in type, so we wouldn't have to define it ourselves ;)

Oh didn't know that but that's even better

> This is something I'm not a fan of though.
> It's a similar issue to session and sessionID on Request. I marked these as non-optional, even though they only exist after the express-session middleware has been used. I did this because those properties will always exist if the module is used properly, sparing the need for unnecessary (and possibly confusing) type assertions.

I am with you here.

> Like you said, it's probably a matter of taste in the end. I'll have to think about this some more before I can make up my mind on what my preferred flavor is.

👍

rbuckton commented:
As an FYI, you can always do something like `export const Session: new (whateverArgsGoHere) => Session;` to make it "class-like" without having to make it an actual `class`.

HoldYourWaffle commented:
> As an FYI, you can always do something like export const Session: new (whateverArgsGoHere) => Session; to make it "class-like" without having to make it an actual class.

That could definitely be another option.

> I'll have to think about this some more before I can make up my mind on what my preferred flavor is.

I have done some thinking in between melting in a heatwave, so there's a good chance this doesn't make any sense, but would something like this work?:

```ts
// Could also use an empty interface here because type-aliases get erased, possibly avoiding confusion
type InitializedSession = Session & SessionData;

interface Request {
    // This union would also be used in all other places of course
    session: Session | InitializedSession;
}
```

As far as I know this would allow easy distinction between an "initialized"- and "uninitialized"-session, because the whole union can be narrowed down with just a single type assertion. This saves the need for having to (needlessly) check each property individually every time, and more 'cleanly' preserves optional/nullable properties on initialized sessions.

However, my brain is most likely affected by the heat right now, so there's a good chance that this doesn't work the way I expect it to.

kstiehl commented:
I kinda like the Idea since it would allow the user to create a custom Type Guard to check whether a session is initialised. 
I think it represents pretty accurate what's going on inside the package. 👍

JLuboff commented:
@HoldYourWaffle Looking forward to this getting merged, thanks for all your effort on this!

HoldYourWaffle commented:
I have pushed another commit with the proposed changes. Based on the (updated) tests it looks like everything is working as it should!

My apologies for the delay due to personal issues. @JLuboff you have no idea how much your comment meant to me...

peterblazejewicz commented on types/express-session/index.d.ts:
```diff
@@ -9,115 +10,354 @@
+     * **Warning:** the default server-side session storage, `MemoryStore`, is purposely not designed for a production environment.
+     * It will leak memory under most conditions, does not scale past a single process, and is only meant for debugging and developing.
+     *
+     * @deprecated
```
from where this `@deprecated` comes?

HoldYourWaffle replied:
I honestly do not remember, I'll remove it

peterblazejewicz commented on types/express-session/index.d.ts:
```diff
@@ -1,6 +1,7 @@
 // Type definitions for express-session 1.17
 // Project: https://github.com/expressjs/session
-// Definitions by: Hiroki Horiuchi <https://github.com/horiuchi>
+// Definitions by: Ravi van Rooijen <https://github.com/HoldYourWaffle>
+//                 Hiroki Horiuchi <https://github.com/horiuchi>
```
```suggestion
// Definitions by: Hiroki Horiuchi <https://github.com/horiuchi>
```

peterblazejewicz commented on types/express-session/index.d.ts:
```diff
@@ -1,6 +1,7 @@
+//                 Hiroki Horiuchi <https://github.com/horiuchi>
 //                 Jacob Bogers <https://github.com/jacobbogers>
 //                 Naoto Yokoyama <https://github.com/builtinnya>
 //                 Ryan Cannon <https://github.com/ry7n>
```
```suggestion
//                 Ryan Cannon <https://github.com/ry7n>
//                 Ravi van Rooijen <https://github.com/HoldYourWaffle>              
```

peterblazejewicz commented on types/passport.socketio/package.json:
```diff
@@ -1,6 +1,6 @@
     "private": true,
     "dependencies": {
-        "connect-mongo": "^3.1.1"
+        "connect-mongo": "HoldYourWaffle/connect-mongo#fix-dt-external-error"
```
what's the reason behind this change?

HoldYourWaffle replied:
See [this comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46576#issuecomment-669489207). It's mostly just a temporary thing to make CI tests pass until I know how to properly solve it.

I'll take a look at your other comments tomorrow.

andrewbranch replied:
The best solution is to remove this dependency from @types/passport.socketio. It’s only used in tests, which has the quite unfortunate side effect of unnecessarily adding connect-mongo to the actual npm dependencies of @types/passport.socketio. If you remove it, you will solve your problem and save the 411 people per week who download @types/passport.socketio from bloating their node_modules 🎉

orta commented:
Poke

HoldYourWaffle commented:
Thanks for the poke @orta, this issue once again slipped from my mind...

@peterblazejewicz I have addressed your last concerns and squashed all commits.
As far as I can see the only issue remaining is the [circular typing-dependency](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46576#issuecomment-669489207). Does anyone have an idea on how that could be solved?

andrewbranch commented:
@peterblazejewicz it looks like you previously reviewed this for correctness, and your suggestions have been addressed—how do you feel about it? I have spot checked it and am happy with it infrastructure-wise, but have not carefully looked at the contents of the typings. If you have, I think we can proceed.

peterblazejewicz commented:
> it looks like you previously reviewed this for correctness, and your suggestions have been addressed—how do you feel about it? I have spot checked it and am happy with it infrastructure-wise, but have not carefully looked at the contents of the typings. If you have, I think we can proceed.

@andrewbranch 
quite a story with this PR, isn't it? Yes, reviewed, thx!

HoldYourWaffle commented:
> quite a story with this PR, isn't it?

You can say that again...

peterblazejewicz reviewed (approved):
LGTM!
@HoldYourWaffle thx and sry for time spent on this!

HoldYourWaffle commented:
Ready to merge

simllll commented:
This breaks a lot on my side,
I cannot really solve the issue as the session is always defined as `Session | InitializedSession`. Therefore typechecker tells me that the property, that I have declared, does not exist on Session. Well it's right, because of the current defintion it only exists on "InitializedSession". If I do an explicit typecast to InitiailzedSession it works, but this seems not the right way.

HoldYourWaffle commented:
First and foremost: I'm sorry that this breaks stuff on your side, but please keep in mind that this was unavoidable.

The reasoning for this choice is described in these comments:
- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46576#issuecomment-670154722
- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46576#issuecomment-670195745
- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46576#issuecomment-670401404
- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46576#issuecomment-671543084
- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46576#issuecomment-671566889

In short, a session is not *always* initialized, and thus does not *always* satisfy the `SessionData` interface. The proper way to deal with this is to use a type-guard that can distinguish between a `InitializedSession` and an (uninitialized) `Session`.
An example for this can be found in the updated test case:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f81282b0c80c7f3963fad54dfe923698a2040c74/types/express-session/express-session-tests.ts#L47-L71

simllll commented:
I saw that but I cannot really follow the reasoning, I understand that the session is not initialized by default. But this is also true for different properties. My session has some properties which are just "initailized" after some speical actions.
So at the beginning my session is empty: {}.. I expect I can check for req.session.userId .. WITHOUT any type errors. Which is not the case right now, but the session itself is there, even if it is empty and not initalized.
Anyway, now the user logs in, the sessio nois now {userId: '1234'}; the check (also in your example) assumes now that Session is intialized and all properties are there, which is not true, as my other propety, let's say it is "lastAction" is not set yet.
So after a while I get another property in there .. let's say it's called "lastAction". Now my session is complete and SessionData has {userId: 1234', lastAction: 'someAction' } in there.

So I would argue that session is actually always there, but properties of the session are "all the time" optional. This is also how express-session is implemented. The session object is there, and only if there are properties defined or changed, the session is saved / created.  (see https://www.npmjs.com/package/express-session#saveuninitialized or https://www.npmjs.com/package/express-session#resave).

To summarize: Checking if the session is "initialized" or not is not really relevant, it is more relevant if a property of the session is available, and therefore Partial<> would be the more appropiate approach in my opinion.

HoldYourWaffle commented:
> So at the beginning my session is empty: {}.. I expect I can check for req.session.userId .. WITHOUT any type errors.

Why do you expect this? TypeScript never allows using keys that don't exist on a type. I think you might be looking for the `in` operator, perhaps [this page in the TS manual](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types) can help.

> the check (also in your example) assumes now that Session is intialized and all properties are there, which is not true, as my other propety, let's say it is "lastAction" is not set yet

This depends entirely on how strict you define your type assertion. In most cases it would be enough to determine that a session is "initialized" if it contains at least one "initialized property" (at least in my experience).

Keep in mind that you can still use optional and nullable properties as usual, which brings me to my next point:

> it is more relevant if a property of the session is available, and therefore Partial<> would be the more appropiate approach in my opinion.

I'm not sure how `Partial` would be the right solution here, because it throws away all information on optional properties, which I mentioned in [this comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46576#issuecomment-670195745).

Does this answer your question? I feel like I haven't quite yet, please feel free to ask more if that's the case.

simllll commented:
Thanks for your effort in explaining this to me ;).

Regarding point 1: I'm not saying that I haven't defined it, I have defined it as optional (or just as regular type) that my session can have a userId. When I check for existance, I expect not getting a type error. Because I have already defined the property in typescript, and because express-session has a "session" object holding for me which is empty by the time of first checking. Which is in my point of view a legit approach of how to handle a situation like this.
The express-session object is an "object", which can have properties, but all these properties can be defined/initialized or not.
That's what I mean by "Partial". All session properties should be optional all the time, as you always have to check if they are there. There could awlays be scenarios where a session is initialized* and the properties are still different to another session of another user.

Let's start by defining "session initialized", as I believe that's the real culprit here. Correct me when I'm wrong, but as far as I understand express-session a session is 
* un-initialized: when the user has no cookie set, and therefore a session has not been "restored" from a previous call. It's just super empty. As long as I don't touch it, it will also not get initialized (=a cookie will be set). But: I can touch it and express-session will initialize it for me (=again,just the cookie will be set, for the programmer itself is not any difference, expect the properties are not initialized)!
* initialized: a cookie (or any other way of putting a session id into express-session) restores a session object from somewhere  (e.g. redis) and sets the properties to the values it has been set in an earlier request.

As a user (=programmer) there is no such state as "initialized" or "uninitialized" of a session, or at least it's not really relevant at all. There are different properties of the session though, which can be initialized or not initialized. 
E.g. I have one guest user, that is not logged in (no userId), but she has a lastAction already set.

Therefore it makes no differnce if it is a first request or the 331th request, the situation is always the same, regardless of there is a session cookie set or not.

Can you follow my reasoning that leads me to the point of view that I think the current implementation is not quite correct?

peterblazejewicz commented:
@simllll 
Without much discussion on nature of the problem, would you just want to be able to use, wihtout much fusss:
```ts
app.use((req, res, next) => {
  if (!!req.session.views) {
    req.session.views++;
    res.setHeader('Content-Type', 'text/html');
    res.write(`<p>views: ${req.session.views}</p>`);
    res.write(`<p>expires in: ${((req.session.cookie.maxAge || 0) / 1000)}s</p>`);
    res.end();
  } else {
    req.session.views = 1;
    res.end('welcome to the session demo. refresh!');
  }
````
(so on first visit, set coutner to 1, then increase). Thx!

simllll commented:
@peterblazejewicz Yes exactly, this is also how I understand the implementation of express-session how to use sessions. Or do I miss something here?

peterblazejewicz commented:
@HoldYourWaffle Ravi?

kstiehl commented:
Since I was one of the commenters that brought up the whole 'initialised not initialised' foobar I feel the need to say that I think @simllll is right. There is no such thing as a fully initialised session. So forcing the user to write typeguards doesn't bring you anything but pain. 

The whole thing wasn't probably one of my brightest moments.

andrewbranch commented:
I think there’s a good argument for making it easy to access data-bag-esque properties on `Session` as optional properties by default. It looks like this solution requires users to write a type assertion or user-defined type guard in order to see their `SessionData` properties at all, but once that is done, it has the advantage that non-optional properties on `SessionData` appear non-optional, which could certainly be correct/desirable depending on how you’re using it.

But you could just as well go from a type like `Session & Partial<SessionData>` to `Session & SessionData` with a type assertion or user-defined type guard, right? So you could a) treat all your data properties as optional, without needing to narrow the type given to you by default, and b) treat all your data properties as non-optional (as declared) with a type guard. Whereas now, it sounds like the situation is you need a type guard for _both situations_, which seems unnecessarily painful.

kstiehl commented:
@andrewbranch I still think `Session & Partial<SessionData>` is technically correct since it literally says nothing is guaranteed but you potentially have the following properties. And I can't see how this would break anything.

simllll commented:
Fully agree!

simllll commented:
I've created a new PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/49479 to fix the mentioned issue, please review :-)

HoldYourWaffle commented:
I'm currently not able to thoroughly catch up on the conversation (& review the PR), I'll probably take a look tonight or maybe tomorrow
________________________________
From: Simon Tretter <notifications@github.com>
Sent: Wednesday, November 11, 2020 11:46:44 AM
To: DefinitelyTyped/DefinitelyTyped <DefinitelyTyped@noreply.github.com>
Cc: Ravi van Rooijen <ravivanrooijen@live.nl>; Mention <mention@noreply.github.com>
Subject: Re: [DefinitelyTyped/DefinitelyTyped] Rewrite (thereby fix) types for express-session (second attempt) (#46576)


I've created a new PR #49479<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FDefinitelyTyped%2FDefinitelyTyped%2Fpull%2F49479&data=04%7C01%7C%7C9ea2958614ba42f500d708d8862f157e%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637406884058885199%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=cM5cfZa2cRtkSov3G2ANxtpV7SYFbWUGXt0PcqnJuR4%3D&reserved=0> to fix the mentioned issue, please review :-)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FDefinitelyTyped%2FDefinitelyTyped%2Fpull%2F46576%23issuecomment-725350488&data=04%7C01%7C%7C9ea2958614ba42f500d708d8862f157e%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637406884058895204%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=KhjWSezgziGVORuftYdbYJyfxXQk%2Fk30%2FXUyfD1VUAU%3D&reserved=0>, or unsubscribe<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACXQ3QDSUE6XSQDQHVVLHYLSPJTRJANCNFSM4PVZ2G3A&data=04%7C01%7C%7C9ea2958614ba42f500d708d8862f157e%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637406884058905195%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=YOfR2aVf2gQJBpNUI87bKYnx%2BKYFj%2FNwQBgOpmEBJUk%3D&reserved=0>.

rchl commented:
I'm having issues extending `SessionData` with those updated definitions.

I believe I'm doing it correctly, by declaring a module in `index.d.ts` and extending the `SessionData` within it:
```ts
/// <reference types="express-session" />

declare module 'express-session' {
    interface SessionData {
        foo?: string;
    }
}
```

Yet, in my actual code, this module declaration overrides the actual module, making the `express-session` export only contain that data.

I would think that referencing the `express-session` as I've done here should make sure that that the module is augmented rather than overridden but it doesn't work for some reason.

Here is a minimal repo reproducing this issue: https://github.com/rchl/express-session-types

HoldYourWaffle commented:
@rchl Thank you for providing a minimal repository for this, it makes my life a lot easier 😄

Changing the triple-slash-directive to an import fixes the issue here. To piggy-back off of your example:

**Declaration file:**
```ts
import 'express-session' // you can also import actual types if you need them

declare module 'express-session' {
    interface SessionData {
        foo?: string;
    }
}
```

**Using file:**
```ts
import { SessionData, SessionOptions } from 'express-session'

const options: SessionOptions = { secret: 'xxx' }
const data: SessionData = {};

if (data.foo) {
	// TS recognized foo
}
```

rchl commented:
Thank you for this solution. But if anyone has some more insights on why referencing is not enough, I'd be glad to hear.

The problem with importing is that it changes semantics as the file is no longer an ambient module and this will break more complicated cases like mine. But it might be doable to split things into separate files to overcome that. But still, I would think that referencing should work and I wonder what makes it not work...

HoldYourWaffle commented:
I did some more investigating, but I can't find a combination of references/imports/exports/declaration that makes this work without changing the semantics of the declaration file.

I'm by no means an expert on TS module vs ambient semantics, but I *think* that this could be due to the way `express-session` exports its interfaces (using a namespace that's exported with `export =`).

Unfortunately I don't think there's a way in which the `express-session`-declarations could be restructured to fix this, but as I said, not an expert 😅

rchl commented:
FYI: My issue is solvable by splitting type definitions into separate files.

andrewbranch commented:
@rchl you have discovered the incredibly confusing difference between an ambient module declaration and a module augmentation. The former overrides a module that exists elsewhere, while the latter merges with it. They look exactly the same, except that an ambient module declaration occurs in a script-scope declaration file, while the latter occurs in a module-scope declaration or TypeScript file. Having any imports or exports at all turns a file into a module, so an ambient module declaration turns into a module augmentation as soon as you add the `import "express-session"`, or indeed `import` literally anything else, or even `export {}`.

imtrent commented:
I'm having issues using these updated definitions. What I am trying to do is add userId on to the req.session. I use to be able to do an intersection on the req like so, but now this no longer works.
```javascript
// context.d.ts
export type Context = {
    req: Request & { session: Express.Session };
    res: Response;
};
```

With these new updates, I have added the userId to the SessionData through declaration merging, but this is still not working.
```javascript
// express-session.d.ts
/// <reference types="express-session" />
declare module 'express-session' {
  interface SessionData {
    userId?: number;
  }
}
```
```javascript
// context.d.ts
import { Request, Response } from 'express';
import { Session } from 'express-session';

export type Context = {
  req: Request & { session: Session };
  res: Response;
};
```


Any ideas on how I can get the userId property on the session and still have access to the other typings of destroy, regenerate, save, reload, etc.?

rchl commented:
@imtrent You seem to be hitting the same issue as me above. Just look at the previous comments.
(But basically, it boils down to doing `import 'express-session'` explicitly instead of referencing types.)

Also, you don't need to extend `Request` with `Express.Session` as it's done already by the express-session types.

imtrent commented:
@rchl Thanks for getting back so quick!

I did see the previous comments and have tried importing it instead, however, I'm still running into the same problems.

Here is what I am working with right now
```javascript
// express-session.d.ts
import 'express-session';

declare module 'express-session' {
  interface SessionData {
    userId?: number;
  }
}
```
```javascript
// context.d.ts
import { Request, Response } from 'express';

export type Context = {
  req: Request;
  res: Response;
};
```
This errors out to `Property 'userId' does not exist on type 'Session & Partial<SessionData>'.`

simllll commented:
What kind of setup do you have? E.g. are you using monorepo or anything that could cause that you have different node_modules / different versions of express-session. This is what causes troubles from time to time in my setup...

rchl commented:
That looks correct to me. I guess make sure that `express-session.d.ts` is being imported somewhere and maybe create a repro project if you are still having issues.

imtrent commented:
@simllll @rchl Looks like the declaration was just not getting picked up when compiling 🤦‍♂️.  I was able to solve this problem by adding typeRoots to my tsconfig. Thanks for the help!

0x2t1ff commented:
I agree with the changes brought on this PR but I'm not so sure on the chosen versioning. Previous working projects that I worked on, now are encountering errors. I had a similar use case than the one that @imtrent brought up. So I'm just curious of why wasn't this treated as a MAJOR bump since it broke backwards compatibility?

rchl commented:
@0x2t1ff I believe the rules for versioning the types are a bit different than for versioning libraries.

The FAQ touches a bit on that but seems like not on the specific case where types have a breaking change: https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-add-a-dom-api-not-present-in-typescript-by-default

But in the end, the version of the types must match the library version in regards to `major.minor` components so you can't just up the major version of types because then it wouldn't match the library version. You'd need to up the version of the library also but since library and types are maintained separately, it's not gonna work like that.

0x2t1ff commented:
@rchl Oh I see, sorry for not reading the rules throughly before commenting, but I thought it would be great to point out that some changes may be breaking so it would be great to have that in mind, since it took me a while to figure out what was happening.

Anyway, thanks for the explanation, and thanks for the work that you guys do :D

izhang05 commented:
@imtrent Can you elaborate a little bit more on how you solved your problem. I think I'm having the same issue. Thanks!

imtrent commented:
@izhang05 Hey! Another Ian! 

Your declaration file should look something like this, however, your interface values might be different than mine based on whatever your use case is.
```javascript
import 'express-session';

declare module 'express-session' {
  interface SessionData {
    userId: number;
  }
}
```

I solved my problem by specifying the type roots in the compiler options of the tsconfig.json file:
```
"typeRoots": ["./src/types", "./node_modules/@types"]
```

I also nested the declaration file in an express-session directory within my types directory. My structure is as follows:
```
- types
    - express-session
        - index.d.ts
```

izhang05 commented:
@imtrent Thank you for getting back! That fixed the issue I was having.

silveoj commented:
I can't fix it for 2 months)

`tsc` command works with 1.17.0 and 1.17.3
but `ts-node` package works only with 1.17.0. 

```
 // ts config
  "typeRoots": [
       "./src/types",
       "./node_modules/@types"
     ]
```
My `types/express-session/index.d.ts`

```
import 'express-session';

import { IMyInt} from '@interfaces/IMyInt';

declare module 'express-session' {
  interface SessionData { // I also tried just Session // I tried export interface ...
    myStr: string;  // Property 'myStr' does not exist on type 'Session & Partial<SessionData>'.
    myProp1?: IMyInt; // Property 'myProp1' does not exist on type 'Session & Partial<SessionData>'.
    // cookie: Cookie;
  }
}
```

doesn't work

HoldYourWaffle commented:
@silveoj This is probably caused by [this optimization in `ts-node`](https://github.com/TypeStrong/ts-node#help-my-types-are-missing).
It should work if your declaration file is explicitly referenced somewhere.

silveoj commented:
@HoldYourWaffle Thanks. Yes, I tried `/// <reference types="express-session" />` and similar versions in different places. It doesn't work. Maybe I missed something.

It works when I added `ts-node` into `tsconfig.json`
```
  "compilerOptions": { ... },
  "ts-node": {
    "files": true
  },
```
That works with and without `typeRoots`

silveoj commented:
@HoldYourWaffle Thanks. Yes, I tried `/// <reference types="express-session" />` and similar versions in different places. It doesn't work. Maybe I missed something.

It works only when I added `ts-node` into `tsconfig.json`
```
  "compilerOptions": { ... },
  "ts-node": {
    "files": true
  },
```
That works with and without `typeRoots`

Helpful comment https://github.com/TypeStrong/ts-node/issues/1132#issuecomment-716642560

HoldYourWaffle commented:
Hmmm, it's been a while since I've dealt with this issue myself. I think I fixed it by using a file-path in my triple-slash-directive instead of a module name, but I'm not sure if that'll work here.

silveoj commented:
Doesn't work)
`<reference types="./types/express-session" />`
`<reference types="./types/express-session/index.d.ts" />`

But in other project with the same structure and the same packages and it versions it works without reference and tsconfig. If I find the reason I will describe it.

HoldYourWaffle commented:
Maybe try using `reference path="..."` instead of `types`?
I can't find the code from when I had the problem, so if this doesn't work I'm out of options 😅

Let me know if you find something though!

silveoj commented:
It didn't work because I have
`"types": []` (empty array or not) inside `compilerOptions`

If I remove `types` and add `typeRoots` it will work.
I don't need:
```
  "ts-node": {
    "files": true
  },
```

silveoj commented:
It didn't work because I have
`"types": []` (empty array or not) inside `compilerOptions`

It will work if:
- add `typeRoots` and remove `types`
- add `typeRoots` and add `"types": ["express-session"]`
I don't need:
```
  "ts-node": {
    "files": true
  },
```

silveoj commented:
It didn't work because I have
`"types": []` (empty array or not) inside `compilerOptions`

It will work if:
- add `typeRoots` and remove (even empty) `types`
- add `typeRoots` and add `"types": ["express-session"]`

i.e. both only or without `types`

I don't need:
```
  "ts-node": {
    "files": true
  },
```

alfonsoprado commented:
I tried the above examples and it didn't work for me. Perhaps the above happened because there were files with definitions that overwritten the previous definition. In the end I did the following and it worked for me 

```
// tsconfig.json
"typeRoots": [
	"@types",
	"./node_modules/@types"
]
```

```
// file: @types/express/index.d.ts
import "express-session";

// what we are all looking to solve xD
// req.session.propertyX
declare module "express-session" {
	interface SessionData {
		propertyX: number;
	}
}

// additional properties on other objects
// req.propertyY
declare global {
	namespace Express {
		interface Request {
			propertyX: number;
		}
	}
}

```
hope it works for you, and saves time.
